File: src/main/java/gargoyle/netsquares/util/InterpolUtils.java
package gargoyle.netsquares.util;

import java.awt.Color;

public final class InterpolUtils {
    private InterpolUtils() {
    }

    public static Color interpolate(Color color1, Color color2, int steps, int step) {
        int red = interpolate(color1.getRed(), color2.getRed(), steps, step);
        int green = interpolate(color1.getGreen(), color2.getGreen(), steps, step);
        int blue = interpolate(color1.getBlue(), color2.getBlue(), steps, step);
        return new Color(red, green, blue);
    }

    private static int interpolate(int value1, int value2, int steps, int step) {
        return (int) (value1 + (value2 - value1) / (double) steps * step);
    }

    public static int toRange(int v, int min, int max) {
        return Math.max(min, Math.min(v, max));
    }
}

File: src/main/kotlin/gargoyle/netsquares/util/InterpolUtils.kt
package gargoyle.netsquares.util

import java.awt.Color

object InterpolUtils {
    fun interpolate(color1: Color, color2: Color, steps: Int, step: Int): Color {
        val red = interpolate(color1.red, color2.red, steps, step)
        val green = interpolate(color1.green, color2.green, steps, step)
        val blue = interpolate(color1.blue, color2.blue, steps, step)
        return Color(red, green, blue)
    }

    private fun interpolate(value1: Int, value2: Int, steps: Int, step: Int): Int {
        return (value1 + (value2 - value1) / steps.toDouble() * step).toInt()
    }

    fun toRange(v: Int, min: Int, max: Int): Int {
        return min.coerceAtLeast(v.coerceAtMost(max))
    }
}
